haystack/modeling/utils.py

```diff
@@ -68,8 +68,12 @@ def initialize_device_settings(
         devices_to_use = [torch.device("cpu")]
         n_gpu = 0
     elif devices:
-        devices_to_use = devices
-        n_gpu = sum(1 for device in devices if "cpu" not in device.type)
+        if isinstance(devices[0], str):
```

Comment:
We could consider replacing this if statement with `any(isinstance(x, str) for x in devices)` which will check if any of the devices provided are of type string. For example, if the user happens to provide `[torch.device('cuda:0'), 'cuda:1']` then we would still construct a list of only `torch.device` types. 

This will work because `torch.device(torch.device('cuda'))` still returns `torch.device('cuda')` so there is no danger there.

Reply:
Great point, thank you @sjrl